File: scripts/day4-part2-simple.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <chrono>

using namespace std;

int countNeighbors(const vector<string>& grid, int row, int col) {
    int count = 0;
    int rows = grid.size();
    int cols = grid[0].size();
    
    // 8 directions
    int dx[] = {-1, -1, -1, 0, 0, 1, 1, 1};
    int dy[] = {-1, 0, 1, -1, 1, -1, 0, 1};
    
    for (int i = 0; i < 8; i++) {
        int newRow = row + dx[i];
        int newCol = col + dy[i];
        if (newRow >= 0 && newRow < rows && newCol >= 0 && newCol < cols) {
            if (grid[newRow][newCol] == '@') {
                count++;
            }
        }
    }
    
    return count;
}

int solve() {
    ifstream file("../inputs/day-4-input.txt");
    vector<string> grid;
    string line;
    
    while (getline(file, line)) {
        grid.push_back(line);
    }
    file.close();
    
    int totalRemoved = 0;
    bool changed = true;
    
    while (changed) {
        changed = false;
        vector<pair<int, int>> toRemove;
        
        // Find all cells with < 4 neighbors
        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid[i].size(); j++) {
                if (grid[i][j] == '@') {
                    int neighbors = countNeighbors(grid, i, j);
                    if (neighbors < 4) {
                        toRemove.push_back({i, j});
                    }
                }
            }
        }
        
        // Remove them
        if (!toRemove.empty()) {
            changed = true;
            totalRemoved += toRemove.size();
            for (auto& pos : toRemove) {
                grid[pos.first][pos.second] = '.';
            }
        }
    }
    
    return totalRemoved;
}

int main() {
    const int iterations = 1;
    int answer = 0;
    
    auto start = chrono::high_resolution_clock::now();
    
    for (int i = 0; i < iterations; i++) {
        answer = solve();
    }
    
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::microseconds>(end - start);
    
    double avgTime = static_cast<double>(duration.count()) / iterations;
    double totalTimeMs = duration.count() / 1000.0;
    
    cout << "C++ (Day 4 Part 2 - Simple)" << endl;
    cout << "Answer: " << answer << endl;
    cout << "Average time: " << avgTime << " microseconds" << endl;
    cout << "Total time: " << totalTimeMs << " ms" << endl;
    
    return 0;
}

File: scripts/day4-part2-simple.js
const fs = require("fs");

function solvePart2(grid) {
  const mutableGrid = grid.map((row) => row.split(""));
  const rows = mutableGrid.length;
  const cols = mutableGrid[0].length;

  const directions = [
    [-1, -1],
    [-1, 0],
    [-1, 1],
    [0, -1],
    [0, 1],
    [1, -1],
    [1, 0],
    [1, 1],
  ];

  let totalRemoved = 0;

  while (true) {
    const accessible = [];

    for (let r = 0; r < rows; r++) {
      for (let c = 0; c < cols; c++) {
        if (mutableGrid[r][c] === "@") {
          let adjacentRolls = 0;

          for (const [dr, dc] of directions) {
            const nr = r + dr;
            const nc = c + dc;

            if (
              nr >= 0 &&
              nr < rows &&
              nc >= 0 &&
              nc < cols &&
              mutableGrid[nr][nc] === "@"
            ) {
              adjacentRolls++;
            }
          }

          if (adjacentRolls < 4) {
            accessible.push([r, c]);
          }
        }
      }
    }

    if (accessible.length === 0) break;

    for (const [r, c] of accessible) {
      mutableGrid[r][c] = ".";
    }

    totalRemoved += accessible.length;
  }

  return totalRemoved;
}

const input = fs
  .readFileSync("../inputs/day-4-input.txt", "utf-8")
  .trim()
  .split(/\r?\n/);

const iterations = 1;
const startTime = process.hrtime.bigint();

let result;
for (let i = 0; i < iterations; i++) {
  result = solvePart2(input);
}

const endTime = process.hrtime.bigint();
const totalNanos = Number(endTime - startTime);
const avgMicros = (totalNanos / iterations / 1000).toFixed(3);

console.log(`JavaScript (Day 4 Part 2 - Simple)`);
console.log(`Answer: ${result}`);
console.log(`Average time: ${avgMicros} microseconds (${iterations} iterations)`);
console.log(`Total time: ${(totalNanos / 1000000).toFixed(2)} ms`);
